net/sock/dtls: allow using multiple credentials

leandrolanzieri opened the pull request:
### Contribution description
Currently only one credential can be associated to each DTLS sock. This limits certain use cases that need to establish multiple connections on a single sock using potentially multiple credentials.

This PR proposes an extension of the sock DTLS API to allow registering multiple credentials to be used on a single sock. An application may register callbacks to specify a certain credential from the list based on the session information (and, when using PSK, based on the hint that may have been sent).

Additionally, now a server can send a PSK Identity hint to a client during the handshake, to help deciding on the PSK Identity that should be used.

Finally, the example application has been extended to show these functionalities.

### Testing procedure
- The example application should be working both when using PSK and ECC (for more information enable the debug messages in the sock_dtls.c file).
- Check that the correct hint is being sent in PSK
- Check that the correct RPK is being used in ECC
- Read the documentation
- A single credential use case should still work seamlessly

### Issues/PRs references
None so far

leandrolanzieri commented on pkg/tinydtls/contrib/sock_dtls.c:
```diff
@@ -162,37 +170,80 @@ static int _get_psk_info(struct dtls_context_t *ctx, const session_t *session,
+            }
+        }
+
+        /* if no callback set or no valid credential returned, try to find a valid registered one */
+        DEBUG("sock_dtls: trying to get first PSK credential\n");
+        for (unsigned i = 0; i < sock->tags_len; i++) {
+            if (credman_get(&credential, sock->tags[i], CREDMAN_TYPE_PSK) == CREDMAN_OK) {
+                c = credential.params.psk.id.s;
+                c_len = credential.params.psk.id.len;
+                break;
+            }
+        }
```
Note to self: prioritize credential by hint.

leandrolanzieri replied:
Done in 01dc4dfd5d454871f054d7188fd24605f01925e2

leandrolanzieri commented:
@janosbrodbeck the new definitions are not split into their own header, this should avoid the issues when using Sock Async.

pokgak commented on pkg/tinydtls/contrib/sock_dtls.c:
```diff
@@ -296,6 +392,67 @@ int sock_dtls_create(sock_dtls_t *sock, sock_udp_t *udp_sock,
+    return 0;
+}
+
+int sock_dtls_add_credential(sock_dtls_t *sock, credman_tag_t tag)
```
I'm thinking if we need to call `sock_dtls_add_credential` for each new `credman_add` anyways, would it be better to include `credman_add` in `sock_dtls_add_credential` also? The parameter list might need to change then though.

janosbrodbeck replied:
It is completely legitimate to use credentials for multiple sockets. Running `credman_add()` every time would mean that every time you reuse a credential, you would first have to find out that the credential already exists.

pokgak commented on pkg/tinydtls/contrib/sock_dtls.c:
```diff
@@ -296,6 +392,67 @@ int sock_dtls_create(sock_dtls_t *sock, sock_udp_t *udp_sock,
+        sock->tags_len++;
+        return 0;
+    }
+    DEBUG("sock_dtls: could not new add credential\n");
```
```suggestion
    DEBUG("sock_dtls: could not add new credential\n");
```
minor

pokgak commented on pkg/tinydtls/contrib/sock_dtls.c:
```diff
@@ -75,6 +75,14 @@ static dtls_handler_t _dtls_handler = {
+ * @brief Array of ECDSA keys to convert from credman credentials and send in
+ *        the callback.
+ */
+static dtls_ecdsa_key_t _ecdsa_keys[CONFIG_DTLS_CREDENTIALS_MAX];
```
If the corresponding credentials are removed with `sock_dtls_remove_credential`, will it also be removed from `_ecdsa_keys`?

leandrolanzieri replied:
Currently not. It should actually not interfere with an ongoing session. Now that you mention this, probably `sock_dtls_session_destroy` should mark as free the previously used structure. Does tinyDTLS expose a pointer to the provided structure? If that's not the case, then probably that state should be kept externally (session <-> ecdsa struct).

leandrolanzieri replied:
In d054da2597a45cb3acc4610d80a6df80293738a7 I modified the logic to keep a pointer to the session during the handshake. I did not find any documentation regarding when the key can be freed, currently this is done when the handshake is complete. Maybe we should check what happens on reconnections.

pokgak reviewed (changes requested):
Looks good overall.

janosbrodbeck reviewed (changes requested):
`tests/pkg_tinydtls_sock_async/dtls-server.c` needs adjustments, because the PSK matching changed. The server in the test does not provide an identity for the credential. Don't know why the ID is missing but now it leads with the current changes to not being able to find a matching credential, since it compares with an empty identity. I suspect that since only one credential was possible before, the only existing key was simply used without identity comparison.

```
static const uint8_t psk_id_0[] = PSK_DEFAULT_IDENTITY;
```
needs to be added to `pkg_tinydtls_sock_async/dtls-server.c`. And the ID then needs to be added to the credential.

janosbrodbeck commented:
Looks good to me. Existing code is still working (except the test, but I'll leave that out) and the API is easy to use. Already have a WIP build for my gcoap-dtls running which supports the new functionality.

However, I am for a small extension in `sock_dtls_create()` as I found in a small issue:

When creating a new DTLS socket it is mandatory to provide a credman tag, it includes `sock_dtls_add_credential()`. That was okay when there could only be one credential per socket. For gcoap I defined a pre-defined tag and the user could add their credential and tag it with the gcoap tag. But now there is no need for that anymore, I would prefer to have the possibility to create a socket with an empty tag list. The user has to create the credential anyway, and I see no reason to make the usage of a pre-defined tag mandatory. I would like to leave the tag creation up to the user.

But the current API does not provide that functionality. For gcoap I would have to create the socket with an dummy tag and then remove it with `sock_dtls_remove_credential()` to get an empty tag list. I would be in favour of changing the API tbh.

We should add an option to allow socket creation with empty credential list, e.g. define a `CREDMAN_NO_TAG` and parse this define in the `sock_dtls_create()` to "no tag".

leandrolanzieri commented:
> `static const uint8_t psk_id_0[] = PSK_DEFAULT_IDENTITY;` needs to be added to `pkg_tinydtls_sock_async/dtls-server.c`. And the ID then needs to be added to the credential.

Added in 26cf8cf9f505576a6b0b073a423a601c996e186a.

> But the current API does not provide that functionality. It should be fairly easy to achieve, since `CREDMAN_TAG_EMPTY` already exists. We should parse this option in `sock_dtls_create()` to allow the creation of a socket with an empty credential list. And should also add this possibility to the documentation.

Added in b051dae589485da83e41c1283d5908399f5db8c3.

janosbrodbeck commented on pkg/tinydtls/contrib/sock_dtls.c:
```diff
@@ -135,6 +148,16 @@ static int _event(struct dtls_context_t *ctx, session_t *session,
     }
+
+    if (IS_ACTIVE(CONFIG_DTLS_ECC) && code == DTLS_EVENT_CONNECTED) {
+        for (unsigned i = 0; i < ARRAY_SIZE(_ecdsa_keys); i++) {
```
`_ecdsa_keys` is not declared when `CONFIG_DTLS_ECC` is not set, because `_ecdsa_keys` is declared in a `#ifdef` block. This leads to compilation errors currently.

leandrolanzieri replied:
Yep, I'll change it to `ifdef`

leandrolanzieri replied:
Done

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -468,6 +479,31 @@
  *
+ * ### Multi-credential handling
+ *
+ * Each sock needs at least one credential tag. Extra credentials can be added
```
This might need updating after the changes to allow `CREDMAN_TAG_EMPTY` on `sock_dtls_create`.

pokgak commented on sys/include/net/sock/dtls.h:
```diff
@@ -468,6 +479,31 @@
+ * registering a callback with @ref sock_dtls_set_client_psk_cb. If no callback is registered the
+ * credential is chosen as follows: if a hint is sent by the server, all credentials registered to
+ * the sock are checked for a matching @ref psk_params_t::hint "hint". A credential is select on
+ * matching hint. If no credential matches the hint or no hint is provided, the first PSK credential
```
ditto

leandrolanzieri replied:
Not sure what exactly needs to change here. I made some adaptions to the doc in 74ea5086deba55b80c490b456d0d84b9f229f5ff. Could you take a look?

pokgak reviewed (changes requested):
Almost done. Just need to update the docs.

janosbrodbeck reviewed (approved):
Besides the documentation updates that pokgak suggested, I am very happy.

Tested it again with `examples/dtls-sock`, `tests/pkg_tinydtls_sock_async` and my gcoap-dtls PR without adaption changes and with some first integration changes. The right credentials are used, and existing single credential use cases are also working without changes.

leandrolanzieri commented:
I added another function to get a read-only array of the registered credentials on a sock.

pokgak reviewed (approved):
You can squash now.

leandrolanzieri commented:
Any maintainer that would like to take a look? @miri64? @cgundogan?

miri64 commented on pkg/tinydtls/include/sock_dtls_types.h:
```diff
@@ -74,9 +75,12 @@ struct sock_dtls {
-    credman_tag_t tag;                      /**< Credential tag of a registered
-                                                (D)TLS credential */
+    char psk_hint[CONFIG_DTLS_PSK_ID_HINT_MAX_SIZE]; /**< PSK Identity hint */
+    credman_tag_t tags[CONFIG_DTLS_CREDENTIALS_MAX]; /**< Tags of the available credentials */
+    unsigned tags_len;                      /**< Number of tags in the list 'tags' */
```
I know that this would probably mean a redesign of the `credman` API so I actually would prefer that to be a follow-up, if that is the case, but: Would it make sense to have `tags` (and also maybe `psk_hint`) be a linked list instead of an array?

leandrolanzieri replied:
> I know that this would probably mean a redesign of the `credman` API so I actually would prefer that to be a follow-up, if that is the case, but: Would it make sense to have `tags` (and also maybe `psk_hint`) be a linked list instead of an array?

Not sure if that would actually affect `credman`'s API. The `psk_hint` is just a string, so it should stay as is. It seems a bit more cumbersome from the user perspective having to allocate each list element to register a new tag.

PeterKietzmann commented on pkg/tinydtls/contrib/sock_dtls.c:
```diff
@@ -217,22 +306,61 @@ static int _get_ecdsa_key(struct dtls_context_t *ctx, const session_t *session,
+    ecdsa_key_assignment_t *key = NULL;
+    for (unsigned i = 0; i < CONFIG_DTLS_CREDENTIALS_MAX; i++) {
+        if (!_ecdsa_keys[i].session) {
+            key = &_ecdsa_keys[i];
```
Please document this. You assign the key if there is a free session "slot"?

leandrolanzieri replied:
Yes, tinyDTLS requires that the structure holding the key is allocated by the callback, so I introduced a pool of them.

leandrolanzieri replied:
Comment added

PeterKietzmann commented on pkg/tinydtls/contrib/sock_dtls.c:
```diff
@@ -296,6 +437,76 @@ int sock_dtls_create(sock_dtls_t *sock, sock_udp_t *udp_sock,
+    if (pos >= 0) {
+        sock->tags_len--;
+        for (; (unsigned)pos < sock->tags_len; pos++) {
+            sock->tags[pos] = sock->tags[pos + 1];
```
Really have to copy the tags?

leandrolanzieri replied:
How to store them if not?

PeterKietzmann commented on sys/include/net/sock/dtls.h:
```diff
@@ -30,15 +30,26 @@
- *   5. Send packet to server @ref sock_dtls_send()
+ *   3. Optionally:
+ *          - add extra credentials @ref sock_dtls_add_credential()
+ *          - when using PSK ciphersuites, set a callback for hint reception and credential
```
Change order of first and second bullet to match the above server operations, or does this break the client logic?

leandrolanzieri replied:
Isn't it matching the server order? 

0. set hint (not available in client)
1. add extra credentials
2. set PSK callback (not available in server)
3. set RPK callback

PeterKietzmann commented on sys/include/net/sock/dtls.h:
```diff
@@ -468,6 +479,33 @@
+ * `net/sock/dtls/creds.h`).
+ *
+ * #### Pre-shared Keys Cipher Suites
+ * In the case of PSK, a server can optionally indicate a hint to help the client to decide which
```
Reference to the RFC/Section?

leandrolanzieri replied:
Reference added

PeterKietzmann commented on sys/include/net/sock/dtls.h:
```diff
@@ -468,6 +479,33 @@
+ * In the case of PSK, a server can optionally indicate a hint to help the client to decide which
+ * PSK Identity to use, using @ref sock_dtls_set_psk_identity_hint. The client application can
+ * decide which credential to use based on the sent hint and/or the session information, by
+ * registering a callback with @ref sock_dtls_set_client_psk_cb. If no callback is registered, or
```
I would change the order of arguments. The client can register the callback and on execution, it can utilize the hint

leandrolanzieri replied:
Sorry, I don't understand, can you suggest the change you imagine here?

PeterKietzmann replied:
What I meant was: The client application does not decide which credential to use by **registering** the callback but by **executing** the callback that has been registered **beforehand**. Never mind, reading this again I think it is still understandable this way and shorter.

PeterKietzmann reviewed (commented):
@leandrolanzieri nice contribution! I've left a first round of comments but didn't test it yet. Overall the code looks very good

leandrolanzieri commented on pkg/tinydtls/contrib/sock_dtls.c:
```diff
@@ -162,37 +187,101 @@ static int _get_psk_info(struct dtls_context_t *ctx, const session_t *session,
+                    memcpy(&first, &credential, sizeof(credman_credential_t));
+                }
+
+                if (desc_len == credential.params.psk.hint.len &&
```
Hmm I think there is a missing break after line 237. If there is no hint provided, and a valid PSK credential is found, then we should look no further.

leandrolanzieri replied:
Fixed

leandrolanzieri commented on pkg/tinydtls/contrib/sock_dtls.c:
```diff
@@ -162,37 +187,101 @@ static int _get_psk_info(struct dtls_context_t *ctx, const session_t *session,
-        c_len = credential.params.psk.key.len;
+
+        if (desc) {
+            DEBUG("sock_dtls: looking key for ID: %.*s\n", desc_len, desc);
```
Changed

PeterKietzmann commented:
@leandrolanzieri thanks again for this nice contribution! I've tested multiple scenarios on native and iotlab-m3, all seem to work fine (see report below). The example README could get some enhancement, though. What to expect from a matching (or non-matching) hint, where to change it for experimentation, how to enable to elliptic curve credentials...

<details>
  <summary>examples/dtls-sock, native, PSK</summary>
  
  ## Client

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlsc fe80::98dd:b3ff:fec3:e399 "ABC"
dtlsc fe80::98dd:b3ff:fec3:e399 "ABC"
sock_dtls: credential added in position 1
sock_dtls: event connect
sock_dtls: psk id request
sock_dtls: requesting the application
From [fe80::98dd:b3ff:fec3:e399%5]:20220
Client got hint: hint
sock_dtls: trying to get first PSK credential
sock_dtls: psk key request
sock_dtls: looking for key for ID: Client_identity
sock_dtls: comparing to tag 2, with ID: Wrong_identity
sock_dtls: comparing to tag 3, with ID: Client_identity
sock_dtls: found
sock_dtls: event connected
Connection to server successful
Sent DTLS message
sock_dtls: decrypted message arrived
Received 3 bytes: "ABC"
Terminating

```

## Server

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlss start
dtlss start
> ifconfig
ifconfig
Iface  5  HWaddr: 9A:DD:B3:C3:E3:99 
          L2-PDU:1500  MTU:1500  HL:64  Source address length: 6
          Link type: wired
          inet6 addr: fe80::98dd:b3ff:fec3:e399  scope: link  VAL
          inet6 group: ff02::1
          inet6 group: ff02::1:ffc3:e399
          
> sock_dtls: error receiving UDP packet: -110
sock_dtls: psk hint request
sock_dtls: psk key request
sock_dtls: looking for key for ID: Client_identity
sock_dtls: comparing to tag 10, with ID: Client_identity
sock_dtls: found
sock_dtls: event connected
New client connected
sock_dtls: decrypted message arrived
Received 3 bytes -- (echo)
```

![dtls](https://user-images.githubusercontent.com/7765855/112831261-fd591c80-9093-11eb-8ac6-67c8a8252e0a.png)


</details>

<details>
  <summary>examples/dtls-sock, native, PSK (w/o hint)</summary>
  
  ## Client

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlsc fe80::98dd:b3ff:fec3:e399 "ABC"
dtlsc fe80::98dd:b3ff:fec3:e399 "ABC"
sock_dtls: credential added in position 1
sock_dtls: event connect
sock_dtls: psk id request
sock_dtls: requesting the application
From [fe80::98dd:b3ff:fec3:e399%5]:20220
sock_dtls: trying to get first PSK credential
sock_dtls: psk key request
sock_dtls: looking for key for ID: Wrong_identity
sock_dtls: comparing to tag 2, with ID: Wrong_identity
sock_dtls: found

```

## Server

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlss start
dtlss start
> ifconfig
ifconfig
Iface  5  HWaddr: 9A:DD:B3:C3:E3:99 
          L2-PDU:1500  MTU:1500  HL:64  Source address length: 6
          Link type: wired
          inet6 addr: fe80::98dd:b3ff:fec3:e399  scope: link  VAL
          inet6 group: ff02::1
          inet6 group: ff02::1:ffc3:e399
          
> sock_dtls: error receiving UDP packet: -110
sock_dtls: psk hint request
sock_dtls: no hint provided
sock_dtls: psk key request
sock_dtls: looking for key for ID: Wrong_identity
sock_dtls: comparing to tag 10, with ID: Client_identity
sock_dtls: invalid credential params for type 2
```
![dtls_nohint](https://user-images.githubusercontent.com/7765855/112832365-7efd7a00-9095-11eb-9e32-a20b4fe2b166.png)


</details>


<details>
  <summary>examples/dtls-sock, native, ECC</summary>
  
  ## Client

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlsc fe80::98dd:b3ff:fec3:e399 "ABC"
dtlsc fe80::98dd:b3ff:fec3:e399 "ABC"
sock_dtls: event connect
sock_dtls: get ECDSA key
sock_dtls: event connected
Connection to server successful
Sent DTLS message
sock_dtls: decrypted message arrived
Received 3 bytes: "ABC"

```

## Server

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlsc fe80::98dd:b3ff:fec3:e399 "ABC"
dtlsc fe80::98dd:b3ff:fec3:e399 "ABC"
sock_dtls: credential added in position 1
sock_dtls: event connect
sock_dtls: psk id request
sock_dtls: requesting the application
From [fe80::98dd:b3ff:fec3:e399%5]:20220
sock_dtls: trying to get first PSK credential
sock_dtls: psk key request
sock_dtls: looking for key for ID: Wrong_identity
sock_dtls: comparing to tag 2, with ID: Wrong_identity
sock_dtls: found

```
![dtls_ecc](https://user-images.githubusercontent.com/7765855/112832784-19f65400-9096-11eb-8155-2e7bd25c705a.png)

</details>

<details>
  <summary>examples/dtls-sock, iotlab-m3, PSK</summary>
  
  ## Client

```
> main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlsc fe80::ec1a:a605:46d1:c2d9 "ABC"
dtlsc fe80::ec1a:a605:46d1:c2d9 "ABC"
sock_dtls: credential added in position 1
sock_dtls: event connect
sock_dtls: psk id request
sock_dtls: requesting the application
From [fe80::ec1a:a605:46d1:c2d9%5]:20220
Client got hint: hint
sock_dtls: trying to get first PSK credential
sock_dtls: psk key request
sock_dtls: looking for key for ID: Client_identity
sock_dtls: comparing to tag 2, with ID: Wrong_identity
sock_dtls: comparing to tag 3, with ID: Client_identity
sock_dtls: found
sock_dtls: event connected
Connection to server successful
Sent DTLS message
sock_dtls: decrypted message arrived
Received 3 bytes: "ABC"
Terminating
```

## Server

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlss start
dtlss start
> ifconfig
ifconfig
Iface  5  HWaddr: 42:D9  Channel: 26  Page: 0  NID: 0x23  PHY: O-QPSK

          Long HWaddr: EE:1A:A6:05:46:D1:C2:D9
           TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4
          AUTOACK  ACK_REQ  CSMA  L2-PDU:102  MTU:1280  HL:64  6LO
          IPHC
          Source address length: 8
          Link type: wireless
          inet6 addr: fe80::ec1a:a605:46d1:c2d9  scope: link  VAL
          inet6 group: ff02::1

> sock_dtls: psk hint request
sock_dtls: psk key request
sock_dtls: looking for key for ID: Client_identity
sock_dtls: comparing to tag 10, with ID: Client_identity
sock_dtls: found
sock_dtls: event connected
New client connected
sock_dtls: decrypted message arrived
Received 3 bytes -- (echo)
sock_dtls: error receiving UDP packet: -116
```

</details>

<details>
  <summary>tests/pkg_tinydtls_sock_async, native, PSK</summary>
  
  ## Client

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlsc fe80::98dd:b3ff:fec3:e399
dtlsc fe80::98dd:b3ff:fec3:e399
usage dtlsc <addr> <data>
> dtlsc fe80::98dd:b3ff:fec3:e399 ""^[[DABC
dtlsc fe80::98dd:b3ff:fec3:e399 "ABC
sock_dtls: event connect
> sock_dtls: psk id request
sock_dtls: trying to get first PSK credential
sock_dtls: psk key request
sock_dtls: looking for key for ID: Client_identity
sock_dtls: comparing to tag 2, with ID: Client_identity
sock_dtls: found
sock_dtls: event connected
Session handshake received
Connection to server successful
Sending data ABC"
Sent DTLS message
Session became ready: [fe80::98dd:b3ff:fec3:e399%7]:20220
DTLS message was sent
sock_dtls: decrypted message arrived
Received 6 bytes: ABC"
Terminating session
```

## Server

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlss start
dtlss start
> ifconfig
ifconfig
Iface  7  HWaddr: 9A:DD:B3:C3:E3:99 
          L2-PDU:1500  MTU:1500  HL:64  Source address length: 6
          Link type: wired
          inet6 addr: fe80::98dd:b3ff:fec3:e399  scope: link  VAL
          inet6 group: ff02::1
          inet6 group: ff02::1:ffc3:e399
          
> sock_dtls: psk hint request
sock_dtls: no hint provided
sock_dtls: psk key request
sock_dtls: looking for key for ID: Client_identity
sock_dtls: comparing to tag 10, with ID: Client_identity
sock_dtls: found
sock_dtls: event connected
Session handshake received
New client connected
Session became ready: [fe80::d0f6:c9ff:fe2a:984a%7]:12345
sock_dtls: decrypted message arrived
Received 6 bytes -- (echo)
Session was destroyed by peer: [fe80::d0f6:c9ff:fe2a:984a%7]:12345
```

</details>

leandrolanzieri commented:
> @leandrolanzieri thanks again for this nice contribution! I've tested multiple scenarios on native and iotlab-m3, all seem to work fine (see report below). The example README could get some enhancement, though. What to expect from a matching (or non-matching) hint, where to change it for experimentation, how to enable to elliptic curve credentials...

@PeterKietzmann thanks for the thorough testing! In 4a153eaec038ef040642bf64d0031f2254034bed I added a couple of sections in the application's README. Could you give it a look?

PeterKietzmann commented:
@leandrolanzieri thanks, looks better! Maybe introduce PSK and RPK as abbreviations. Mention that the hint (string) could also be changed, you could also explain that it is printed on the client (for verification by the user). I leave this up to you... Please squash directly

leandrolanzieri commented:
> @leandrolanzieri thanks, looks better! Maybe introduce PSK and RPK as abbreviations. Mention that the hint (string) could also be changed, you could also explain that it is printed on the client (for verification by the user). I leave this up to you... Please squash directly

Added the suggestions and squashed directly!

PeterKietzmann commented on sys/include/net/sock/dtls/creds.h:
```diff
@@ -0,0 +1,152 @@
+ * @retval 0 on success
+ * @retval -1 on error
+ */
+int sock_dtls_set_psk_identity_hint(sock_dtls_t *sock, const char *hint);
```
This is a function for the server only, right? Would it be reasonable to add this to the function name? Like you did for the client functions (for example sock_dtls_set_client_psk_cb)

leandrolanzieri replied:
I renamed it to `sock_dtls_set_server_psk_id_hint`

PeterKietzmann reviewed (approved):
ACK from my side. Anyone else who wants to give this PR a look before merge?

PeterKietzmann commented:
Well then, go!
